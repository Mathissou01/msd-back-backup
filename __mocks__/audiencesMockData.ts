import { GetAudiencesByContractIdDocument } from "../src/graphql/codegen/generated-types";

export const getNoAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [],
        },
      },
    },
  },
];

export const getSingleAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getTwoAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getAllAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
            {
              id: 3,
              attributes: {
                isActive: true,
                type: "Professionnels",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];
