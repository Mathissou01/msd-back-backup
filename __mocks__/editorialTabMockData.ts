import { GetEditoBlockTabDocument } from "../src/graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetEditoBlockTabDocument,
      variables: {
        contractId: "0",
        audienceId: "",
      },
    },
    result: {
      data: {
        getEditoBlockDTO: {
          id: "2",
          displayBlock: true,
          titleContent: "Titre modifié",
          editoContents: [
            {
              id: "7",
              contentType: "tip",
              typeName: "Astuces",
              attributes: {
                title:
                  "Remplacez vos sacs en plastique par des sacs réutilisables, un caddie ou des paniers.",
                status: "published",
                publishedDate: "2022-11-18T10:17:11.158Z",
              },
            },
            {
              id: "14",
              contentType: "freeContent",
              typeName: "Valoriser mes déchets",
              attributes: {
                title: "freecontent test",
                status: "published",
                publishedDate: "2022-12-05T15:59:38.467Z",
              },
            },
            {
              id: "2",
              contentType: "quiz",
              typeName: "Quiz",
              attributes: {
                title: "Nom du quiz",
                status: "published",
                publishedDate: "2022-11-20T02:34:14.535Z",
              },
            },
          ],
        },
        getEditoContentDTOs: [
          {
            id: "1",
            contentType: "tip",
            typeName: "Astuces",
            attributes: {
              title:
                "Découper vos peaux de bananes avant de les jeter dans votre compost.",
              status: "published",
              publishedDate: "2022-11-19T00:14:08.232Z",
            },
          },
          {
            id: "2",
            contentType: "quiz",
            typeName: "Quiz",
            attributes: {
              title: "Nom du quiz",
              status: "published",
              publishedDate: "2022-11-20T02:34:14.535Z",
            },
          },
          {
            id: "7",
            contentType: "tip",
            typeName: "Astuces",
            attributes: {
              title:
                "Remplacez vos sacs en plastique par des sacs réutilisables, un caddie ou des paniers.",
              status: "published",
              publishedDate: "2022-11-18T10:17:11.158Z",
            },
          },
          {
            id: "8",
            contentType: "tip",
            typeName: "Astuces",
            attributes: {
              title: "Titre de l'astuce 3",
              status: "published",
              publishedDate: "2022-11-19T00:14:20.490Z",
            },
          },
          {
            id: "9",
            contentType: "tip",
            typeName: "Astuces",
            attributes: {
              title: "Titre de l'astuce 4",
              status: "published",
              publishedDate: "2022-11-19T00:14:30.273Z",
            },
          },
          {
            id: "11",
            contentType: "quiz",
            typeName: "Quiz",
            attributes: {
              title: "truc",
              status: "published",
              publishedDate: "2022-11-18T10:16:08.936Z",
            },
          },
          {
            id: "12",
            contentType: "news",
            typeName: "Actualités",
            attributes: {
              title: "news test",
              status: "published",
              publishedDate: "2022-12-05T15:59:19.503Z",
            },
          },
          {
            id: "13",
            contentType: "event",
            typeName: "Evénement",
            attributes: {
              title: "event test",
              status: "published",
              publishedDate: "2022-12-05T15:58:52.482Z",
            },
          },
          {
            id: "14",
            contentType: "freeContent",
            typeName: "Valoriser mes déchets",
            attributes: {
              title: "freecontent test",
              status: "published",
              publishedDate: "2022-12-05T15:59:38.467Z",
            },
          },
        ],
      },
    },
  },
];
