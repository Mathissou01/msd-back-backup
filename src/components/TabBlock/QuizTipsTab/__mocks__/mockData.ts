import { GetQuizAndTipsBlockDocument } from "../../../../graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    result: {
      data: {
        contractCustomizations: {
          data: [
            {
              attributes: {
                homepage: {
                  data: {
                    attributes: {
                      quizAndTipsBlock: {
                        data: {
                          id: "1",
                          attributes: {
                            title: "Titre modifi√©",
                            displayBlock: false,
                            displayQuiz: false,
                            quiz: {
                              data: {
                                id: "2",
                                attributes: {
                                  title: "Nom du quiz",
                                  publishedAt: "2022-11-20T02:34:14.535Z",
                                },
                              },
                            },
                            displayTips: false,
                            tips: {
                              data: [],
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          ],
        },
        services: {
          data: [
            {
              attributes: {
                serviceInstance: [
                  {
                    __typename: "ComponentMsdEditorial",
                    id: "4",
                    editorialServices: {
                      data: [
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoQuizzesSubService",
                                isActivated: true,
                                quizzes: {
                                  data: [
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Nom du quiz",
                                        publishedAt: "2022-11-20T02:34:14.535Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "truc",
                                        publishedAt: "2022-11-18T10:16:08.936Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoTipsSubService",
                                isActivated: true,
                                tips: {
                                  data: [
                                    {
                                      id: "5",
                                      attributes: {
                                        title: "nah",
                                        publishedAt: "2022-11-19T00:14:37.403Z",
                                      },
                                    },
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Titre de l'astuce 1",
                                        publishedAt: "2022-11-19T00:14:08.232Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "Titre de l'astuce 2",
                                        publishedAt: "2022-11-18T10:17:11.158Z",
                                      },
                                    },
                                    {
                                      id: "3",
                                      attributes: {
                                        title: "Titre de l'astuce 3",
                                        publishedAt: "2022-11-19T00:14:20.490Z",
                                      },
                                    },
                                    {
                                      id: "4",
                                      attributes: {
                                        title: "Titre de l'astuce 4",
                                        publishedAt: "2022-11-19T00:14:30.273Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    error: new Error("An error occured"),
  },
];
