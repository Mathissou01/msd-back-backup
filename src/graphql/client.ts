import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache({
      typePolicies: {
        UploadFolder: {
          merge: true,
        },
        UploadFile: {
          merge: true,
        },
        Contract: {
          merge: true,
        },
        ContractCustomization: {
          merge: true,
        },
        New: {
          merge: true,
        },
        Tip: {
          merge: true,
        },
        RecyclingGuideService: {
          merge: true,
        },
        WasteForm: {
          merge: true,
        },
        WasteFamily: {
          merge: true,
        },
        Flow: {
          merge: true,
        },
        FlowColor: {
          merge: true,
        },
        DropOffMap: {
          merge: true,
        },
        Request: {
          merge: true,
        },
        RequestAggregate: {
          merge: true,
        },
      },
    }),
  });
}

export default client;
