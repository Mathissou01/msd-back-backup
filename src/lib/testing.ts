import { prettyDOM } from "@testing-library/react";

export function removeUnstableHtmlPropertiesForTesting(
  htmlElement: HTMLElement,
  attributeName: string,
) {
  const domHTML = prettyDOM(htmlElement, Infinity).toString();
  if (!domHTML) return undefined;
  const regex = new RegExp(`${attributeName}(.*)"(.*)"`, "g");
  return domHTML.replaceAll(regex, "");
}
