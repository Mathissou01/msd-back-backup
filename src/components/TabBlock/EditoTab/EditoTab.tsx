import { parseJSON } from "date-fns";
import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, { ReactNode, useEffect, useState } from "react";
import {
  EditoContentDto,
  Enum_Editocontentdto_Status,
  GetEditoBlockTabDocument,
  useGetEditoBlockTabQuery,
  useUpdateEditoBlockTabMutation,
} from "../../../graphql/codegen/generated-types";
import {
  comparePropertyValueByPriority,
  formatDate,
  removeNulls,
} from "../../../lib/utilities";
import { extractEditoBlock } from "../../../lib/graphql-data";
import { useContract } from "../../../hooks/useContract";
import { useFocusFirstElement } from "../../../hooks/useFocusFirstElement";
import CommonLoader from "../../Common/CommonLoader/CommonLoader";
import CommonButton from "../../Common/CommonButton/CommonButton";
import FormInput from "../../Form/FormInput/FormInput";
import FormCheckbox from "../../Form/FormCheckbox/FormCheckbox";
import FormModalButtonInput from "../../Form/FormModalButtonInput/FormModalButtonInput";
import FormMultiselect, {
  IOptionWrapper,
  mapOptionsInWrappers,
} from "../../Form/FormMultiselect/FormMultiselect";
import "./edito-tab.scss";

interface IEditoBlock {
  id: string;
  titleContent: string;
  displayBlock: boolean;
  editoContents?: Array<EditoContentDto | null> | null;
}

export default function EditoTab() {
  /* Static Data */
  const formLabels = {
    title: "Bloc Édito",
    displayBlock: "Afficher ce bloc",
    titleContent: "Titre du bloc",
    editoContents: "Contenus à afficher",
    editoContentsButton: "Sélectionner les articles à afficher",
    editoContentsModal: "Sélection des articles",
    editoContentsModalType: "Article",
  };
  const submitButtonLabel = "Enregistrer les modifications";
  const cancelButtonLabel = "Annuler les modifications";

  /* Methods */
  function editoContentsDisplayTransformFunction(
    editoContent: Array<EditoContentDto | undefined>,
  ): ReactNode {
    return editoContent?.map((editoContent, index) => {
      if (editoContent && editoContent.id) {
        return (
          <p key={editoContent.id + index}>
            {`${editoContent.attributes?.title} - ${formatDate(
              parseJSON(editoContent.attributes?.publishedDate),
            )}`}
          </p>
        );
      }
    });
  }

  function editoContentSelectDisplayTransformFunction(
    editoContent: EditoContentDto,
  ): string {
    return `${editoContent.attributes?.title} - ${formatDate(
      parseJSON(editoContent.attributes?.publishedDate),
    )}`;
  }

  function onContentModalSubmit(submitData: {
    [key: string]: Array<EditoContentDto | undefined>;
  }) {
    const contents = Object.values(submitData)?.filter(removeNulls);
    setValue("editoContents", contents, { shouldDirty: true });
  }

  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        updateEditoBlockId: submitData["id"],
        data: {
          displayBlock: submitData["displayBlock"],
          titleContent: submitData["titleContent"],
          editoContents:
            submitData["editoContents"]?.map(
              (editoContent: EditoContentDto) => editoContent?.id ?? null,
            ) ?? null,
        },
      };
      return updateEditoBlock({
        variables,
        refetchQueries: [
          {
            query: GetEditoBlockTabDocument,
            variables: {
              contractId,
              status: Enum_Editocontentdto_Status.Published,
            },
          },
          "getEditoBlockTab",
        ],
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const { loading, error, data } = useGetEditoBlockTabQuery({
    variables: {
      contractId,
      status: Enum_Editocontentdto_Status.Published,
    },
  });
  const [updateEditoBlock, { loading: mutationLoading, error: mutationError }] =
    useUpdateEditoBlockTabMutation();

  /* Local Data */
  const [editoData, setEditoData] = useState<IEditoBlock>();
  const [editoContents, setEditoContents] = useState<
    Array<IOptionWrapper<EditoContentDto>>
  >([]);
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, setValue, watch, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (data?.getEditoBlockDTO && data.getEditoContentDTOs) {
      const { editoBlock, editoContents } = extractEditoBlock(data);
      if (editoBlock?.id) {
        const mappedData: IEditoBlock = {
          id: editoBlock.id,
          titleContent: editoBlock.titleContent,
          displayBlock: editoBlock.displayBlock,
          editoContents: editoBlock.editoContents,
        };
        setEditoData(mappedData);
        form.reset(mappedData);
      }

      const sortedEditoContents = [...(editoContents ?? [])];
      sortedEditoContents?.sort(
        comparePropertyValueByPriority("contentType", {
          news: 0,
          event: 1,
          tip: 2,
          quiz: 3,
          freeContent: 4,
        }),
      );
      setEditoContents(mapOptionsInWrappers(sortedEditoContents, "typeName"));
    }
  }, [form, data]);

  return (
    <div
      className="c-EditoTab"
      id="panel-edito"
      role="tabpanel"
      aria-labelledby="tab-edito"
    >
      <CommonLoader
        isLoading={loading || isSubmitting || mutationLoading}
        isShowingContent={isSubmitting || mutationLoading}
        hasDelay={isSubmitting || mutationLoading}
        errors={[error, mutationError]}
      >
        <h2 className="c-EditoTab__Title">{formLabels.title}</h2>
        <FormProvider {...form}>
          <form
            className="c-EditoTab__Form"
            onSubmit={handleSubmit(onSubmitValid)}
            ref={useFocusFirstElement()}
          >
            <div className="c-EditoTab__Group c-EditoTab__Group_short">
              <FormCheckbox
                name="displayBlock"
                label={formLabels.displayBlock}
              />
              <FormInput
                type="text"
                name="titleContent"
                label={formLabels.titleContent}
                isRequired={true}
                isDisabled={mutationLoading}
                defaultValue={editoData?.titleContent}
              />
            </div>
            <div className="c-EditoTab__Group">
              <FormModalButtonInput<Array<EditoContentDto>>
                name="editoContents"
                label={formLabels.editoContents}
                displayTransform={editoContentsDisplayTransformFunction}
                buttonLabel={formLabels.editoContentsButton}
                modalTitle={formLabels.editoContentsModal}
                onModalSubmit={onContentModalSubmit}
                isDisabled={mutationLoading}
                isRequired={true}
              >
                <FormMultiselect<EditoContentDto>
                  name="editoContentsSelect"
                  label={formLabels.editoContentsModalType}
                  displayTransform={editoContentSelectDisplayTransformFunction}
                  options={editoContents}
                  selectAmount={3}
                  optionKey={"id"}
                  defaultValues={watch("editoContents")}
                />
              </FormModalButtonInput>
            </div>
            <div className="c-EditoTab__Buttons">
              <CommonButton
                type="submit"
                label={submitButtonLabel}
                style="primary"
                picto="check"
                isDisabled={!isDirty}
              />
              <CommonButton
                type="button"
                label={cancelButtonLabel}
                picto="cross"
                onClick={onCancel}
                isDisabled={!isDirty}
              />
            </div>
          </form>
        </FormProvider>
      </CommonLoader>
    </div>
  );
}
