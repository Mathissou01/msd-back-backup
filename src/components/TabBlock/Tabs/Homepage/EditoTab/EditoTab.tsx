import { parseJSON } from "date-fns";
import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, { ReactNode, useEffect, useState } from "react";
import {
  AudienceEntity,
  EditoContentDto,
  useGetEditoBlockTabByContractIdAndAudienceIdQuery,
  useUpdateEditoBlockByIdMutation,
} from "../../../../../graphql/codegen/generated-types";
import {
  comparePropertyValueByPriority,
  formatDate,
  removeNulls,
} from "../../../../../lib/utilities";
import { extractEditoBlock } from "../../../../../lib/graphql-data";
import { TEditorialContentTypes } from "../../../../../lib/editorial";
import { useContract } from "../../../../../hooks/useContract";
import { useFocusFirstElement } from "../../../../../hooks/useFocusFirstElement";
import CommonLoader from "../../../../Common/CommonLoader/CommonLoader";
import CommonButton from "../../../../Common/CommonButton/CommonButton";
import FormInput from "../../../../Form/FormInput/FormInput";
import FormCheckbox from "../../../../Form/FormCheckbox/FormCheckbox";
import FormModalButtonInput from "../../../../Form/FormModalButtonInput/FormModalButtonInput";
import FormMultiselect, {
  IOptionWrapper,
  mapOptionsInWrappers,
} from "../../../../Form/FormMultiselect/FormMultiselect";
import "./edito-tab.scss";

interface IEditoBlock {
  id: string;
  titleContent: string;
  displayBlock: boolean;
  editoContents?: Array<EditoContentDto | null> | null;
}

interface IEditoTabProps {
  activatedTypes: Array<TEditorialContentTypes>;
  audience?: AudienceEntity;
}

interface IEditoContentsModalFields {
  editoContentsSelect_0: EditoContentDto;
  editoContentsSelect_1: EditoContentDto;
  editoContentsSelect_2: EditoContentDto;
}
export default function EditoTab({ activatedTypes, audience }: IEditoTabProps) {
  /* Static Data */
  const formLabels = {
    title: "Bloc Édito",
    displayBlock: "Afficher ce bloc",
    titleContent: "Titre du bloc",
    editoContents: "Contenus à afficher",
    editoContentsButton: "Sélectionner les articles à afficher",
    editoContentsModal: "Sélection des articles",
    editoContentsModalType: "Article",
  };
  const submitButtonLabel = "Enregistrer les modifications";
  const cancelButtonLabel = "Annuler les modifications";

  /* Methods */
  function editoContentsDisplayTransformFunction(
    editoContent?: Array<EditoContentDto>,
  ): ReactNode {
    return editoContent?.map((editoContent, index) => {
      if (editoContent && editoContent.id) {
        return (
          <p key={editoContent.id + index}>
            {`${editoContent.attributes?.title} - ${formatDate(
              parseJSON(editoContent.attributes?.publishedDate),
            )}`}
          </p>
        );
      }
    });
  }

  function onContentModalSubmit(submitData: IEditoContentsModalFields) {
    return [
      submitData.editoContentsSelect_0,
      submitData.editoContentsSelect_1,
      submitData.editoContentsSelect_2,
    ];
  }

  function editoContentSelectDisplayTransformFunction(
    editoContent: EditoContentDto,
  ): string {
    return `${editoContent.attributes?.title} - ${formatDate(
      parseJSON(editoContent.attributes?.publishedDate),
    )}`;
  }

  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        updateEditoBlockId: submitData["id"],
        data: {
          displayBlock: submitData["displayBlock"],
          titleContent: submitData["titleContent"],
          editoContents:
            submitData["editoContents"]
              .filter(removeNulls)
              ?.map((editoContent: EditoContentDto, index: number) => {
                return {
                  id: editoData?.editoContents
                    ? editoData.editoContents[index]?.componentId
                    : undefined,
                  __typename: "ComponentLinksEditoContent",
                  new:
                    editoContent.contentType === "new" ? editoContent.id : null,
                  event:
                    editoContent.contentType === "event"
                      ? editoContent.id
                      : null,
                  tip:
                    editoContent.contentType === "tip" ? editoContent.id : null,
                  quiz:
                    editoContent.contentType === "quiz"
                      ? editoContent.id
                      : null,
                  freeContent:
                    editoContent.contentType === "free-content"
                      ? editoContent.id
                      : null,
                };
              })
              .filter(removeNulls) ?? null,
        },
      };
      return updateEditoBlock({
        variables,
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const getEditoBlockTabVariables: {
    contractId: string;
    audienceId: string;
  } = {
    contractId,
    audienceId: "",
  };
  if (audience && audience.id) {
    getEditoBlockTabVariables.audienceId = audience.id;
  }
  const { loading, error, data } =
    useGetEditoBlockTabByContractIdAndAudienceIdQuery({
      variables: getEditoBlockTabVariables,
      fetchPolicy: "no-cache",
    });
  const [updateEditoBlock, { loading: mutationLoading, error: mutationError }] =
    useUpdateEditoBlockByIdMutation({
      refetchQueries: ["getEditoBlockTabByContractIdAndAudienceId"],
      awaitRefetchQueries: true,
    });

  /* Local Data */
  const [isInitialized, setIsInitialized] = useState(false);
  const [editoData, setEditoData] = useState<IEditoBlock>();
  const [editoContents, setEditoContents] = useState<
    Array<IOptionWrapper<EditoContentDto>>
  >([]);
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, watch, formState } = form;
  const { isDirty, isSubmitting } = formState;
  useEffect(() => {
    if (data?.getEditoBlockDTO && data.getEditoContentDTOs) {
      const { editoBlock, editoContents } = extractEditoBlock(data);
      if (editoBlock?.id) {
        const mappedData: IEditoBlock = {
          id: editoBlock.id,
          titleContent: editoBlock.titleContent,
          displayBlock: editoBlock.displayBlock,
          editoContents: editoBlock.editoContents,
        };
        setEditoData(mappedData);
        setIsInitialized(true);
        form.reset(mappedData);
      }
      const filteredEditoContents = [...(editoContents ?? [])].filter(
        (content) => {
          if (
            content?.contentType &&
            // typeof TEditoContentTypes &&
            activatedTypes?.includes(
              content.contentType as TEditorialContentTypes,
            )
          ) {
            return content;
          }
        },
      );
      filteredEditoContents?.sort(
        comparePropertyValueByPriority("contentType", {
          new: 0,
          event: 1,
          tip: 2,
          quiz: 3,
          "free-content": 4,
        }),
      );
      setEditoContents(mapOptionsInWrappers(filteredEditoContents, "typeName"));
    }
  }, [form, data, activatedTypes]);

  return (
    <div
      className="c-EditoTab"
      id="panel-edito"
      role="tabpanel"
      aria-labelledby="tab-edito"
    >
      <CommonLoader
        isLoading={loading || isSubmitting || mutationLoading || !isInitialized}
        isShowingContent={isSubmitting || mutationLoading}
        hasDelay={isSubmitting || mutationLoading}
        errors={[error, mutationError]}
      >
        <h2 className="c-EditoTab__Title">{formLabels.title}</h2>
        <FormProvider {...form}>
          <form
            className="c-EditoTab__Form"
            onSubmit={handleSubmit(onSubmitValid)}
            ref={useFocusFirstElement()}
          >
            <div className="c-EditoTab__Group c-EditoTab__Group_short">
              <FormCheckbox
                name="displayBlock"
                label={formLabels.displayBlock}
              />
              <FormInput
                type="text"
                name="titleContent"
                label={formLabels.titleContent}
                isRequired={true}
                isDisabled={mutationLoading}
                defaultValue={editoData?.titleContent}
              />
            </div>
            <div className="c-EditoTab__Group">
              <FormModalButtonInput<
                Array<EditoContentDto>,
                IEditoContentsModalFields
              >
                name="editoContents"
                label={formLabels.editoContents}
                displayTransform={editoContentsDisplayTransformFunction}
                buttonLabel={formLabels.editoContentsButton}
                modalTitle={formLabels.editoContentsModal}
                onModalSubmit={onContentModalSubmit}
                isDisabled={mutationLoading}
                isRequired={true}
              >
                <FormMultiselect<EditoContentDto>
                  name="editoContentsSelect"
                  label={formLabels.editoContentsModalType}
                  displayTransform={editoContentSelectDisplayTransformFunction}
                  options={editoContents}
                  selectAmount={3}
                  optionKey={"uniqueId"}
                  defaultValues={watch("editoContents")}
                />
              </FormModalButtonInput>
            </div>
            <div className="c-EditoTab__Buttons">
              <CommonButton
                type="submit"
                label={submitButtonLabel}
                style="primary"
                picto="check"
                isDisabled={!isDirty}
              />
              <CommonButton
                type="button"
                label={cancelButtonLabel}
                picto="cross"
                onClick={onCancel}
                isDisabled={!isDirty}
              />
            </div>
          </form>
        </FormProvider>
      </CommonLoader>
    </div>
  );
}
