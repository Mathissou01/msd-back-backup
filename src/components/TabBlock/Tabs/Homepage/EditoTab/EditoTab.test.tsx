import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import React from "react";
import { defaultMockData } from "../../../../../../__mocks__/editorialTabMockData";
import EditoTab from "./EditoTab";

describe("EditoTab", () => {
  it("renders loading and loaded state", async () => {
    const { container } = render(
      <MockedProvider mocks={defaultMockData} addTypename={false}>
        <EditoTab
          activatedTypes={["tip", "quiz", "event", "new", "free-content"]}
        />
      </MockedProvider>,
    );

    expect(await screen.findByTestId("common-spinner")).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
    expect(
      await screen.findByText("Nom du quiz - 20/11/2022"),
    ).toBeInTheDocument();
    expect(
      await screen.findByDisplayValue("Titre modifi√©"),
    ).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
  });
});
