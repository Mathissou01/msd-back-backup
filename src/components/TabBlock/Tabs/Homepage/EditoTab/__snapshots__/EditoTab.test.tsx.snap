// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditoTab renders loading and loaded state 1`] = `
<div>
  <div
    aria-labelledby="tab-edito"
    class="c-EditoTab"
    id="panel-edito"
    role="tabpanel"
  >
    <div
      class="c-CommonLoader c-CommonLoader_cover c-CommonLoader_grow"
    >
      <div
        class="c-CommonSpinner"
        data-testid="common-spinner"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`EditoTab renders loading and loaded state 2`] = `
<div>
  <div
    aria-labelledby="tab-edito"
    class="c-EditoTab"
    id="panel-edito"
    role="tabpanel"
  >
    <h2
      class="c-EditoTab__Title"
    >
      Bloc Édito
    </h2>
    <form
      class="c-EditoTab__Form"
    >
      <div
        class="c-EditoTab__Group c-EditoTab__Group_short"
      >
        <div
          class="c-FormCheckbox"
        >
          <input
            class="c-FormCheckbox__Input c-FormCheckbox__Input_checked"
            data-testid="form-checkbox"
            id="displayBlock"
            name="displayBlock"
            type="checkbox"
          />
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="displayBlock"
            >
              <span
                class="c-FormLabel__Label"
              >
                Afficher ce bloc
              </span>
            </label>
          </div>
        </div>
        <div
          class="c-FormInput"
        >
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="titleContent"
            >
              <span
                class="c-FormLabel__Label"
              >
                Titre du bloc *
              </span>
            </label>
            <div
              class="c-FormLabel__Content"
            >
              <input
                aria-errormessage="titleContent_error"
                aria-invalid="false"
                class="c-FormInput__Input"
                data-testid="form-input"
                id="titleContent"
                name="titleContent"
                type="text"
                value="Titre modifié"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="c-EditoTab__Group"
      >
        <div
          class="c-FormModalButtonInput"
          data-testid="form-modal-input"
        >
          <div
            class="c-FormLabel"
          >
            <label
              class="c-FormLabel__LabelWrapper"
              for="editoContents_button"
            >
              <span
                class="c-FormLabel__Label"
              >
                Contenus à afficher *
              </span>
            </label>
          </div>
          <input
            disabled=""
            name="editoContents"
            type="hidden"
            value="[object Object],[object Object],[object Object]"
          />
          <div
            class="c-FormModalButtonInput__Container"
          >
            <p>
              Remplacez vos sacs en plastique par des sacs réutilisables, un caddie ou des paniers. - 18/11/2022
            </p>
            <p>
              freecontent test - 05/12/2022
            </p>
            <p>
              Nom du quiz - 20/11/2022
            </p>
            <div
              aria-errormessage="editoContents_error"
              aria-invalid="false"
            >
              <button
                class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
                id="editoContents_button"
                type="button"
              >
                Sélectionner les articles à afficher
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c-EditoTab__Buttons"
      >
        <button
          class="c-CommonButton c-CommonButton_disabled c-CommonButton_primary c-CommonButton_fontSmall"
          disabled=""
          type="submit"
        >
          <div
            class="c-CommonButton__Picto c-CommonButton__Picto_check"
          />
          Enregistrer les modifications
        </button>
        <button
          class="c-CommonButton c-CommonButton_disabled c-CommonButton_secondary c-CommonButton_fontSmall"
          disabled=""
          type="button"
        >
          <div
            class="c-CommonButton__Picto c-CommonButton__Picto_cross"
          />
          Annuler les modifications
        </button>
      </div>
    </form>
  </div>
</div>
`;
